contracts/consensus/Consensus.sol

```diff
@@ -317,144 +301,232 @@ contract Consensus is MasterCopyNonUpgradable, CoreLifetimeEnum, MosaicVersion,
+            block.number > committeeFormationBlockHeight,
+            "Committee formation height has not yet come to pass."
         );
 
-        uint256 segmentHeight = precommit.committeeFormationBlockHeight;
-        bytes32[] memory seedGenerator = new bytes32[](uint256(COMMITTEE_FORMATION_LENGTH));
-        for (uint256 i = 0; i < COMMITTEE_FORMATION_LENGTH; i = i.add(1)) {
-            seedGenerator[i] = blockhash(segmentHeight);
-            segmentHeight = segmentHeight.sub(1);
-        }
-
-        bytes32 seed = keccak256(
-            abi.encodePacked(seedGenerator)
+        bytes32 seed = hashBlockSegment(
+            committeeFormationBlockHeight,
+            COMMITTEE_FORMATION_LENGTH
         );
 
-        startCommittee(seed, precommit.proposal);
+        currentMetablock.round = MetablockRound.CommitteeFormed;
+        currentMetablock.roundBlockNumber = block.number;
+
+        startCommittee(_metachainId, seed, currentMetablock.metablockHash);
     }
 
     /**
      * @notice Enters a validator into a committee.
      *
      * @dev Function requires:
-     *          - the committee exists
-     *          - the validator is active
-     * 			- the validator is not slashed
+     *          - the given metachain id is not 0
+     *          - the given core's address is not 0
+     *          - the given validator's address is not 0
+     *          - the given 'further' validator's address is not 0
+     *          - the given core is running
+     *          - valid validator is given
+     *              - is active in the given core
+     *              - has not been slashed
+     *          - the corresponding metablock's round is 'CommitteeFormed'
      *
-     * @param _committeeAddress Committee address that validator wants to enter.
+     * @param _metachainId Metachain id for the committee to enter.
      * @param _validator Validator address to enter.
      * @param _furtherMember Validator address that is further member
      *                       compared to the `_validator` address
      */
     function enterCommittee(
-        address _committeeAddress,
+        bytes32 _metachainId,
+        CoreI _core,
         address _validator,
         address _furtherMember
     )
         external
     {
         require(
-            committees[_committeeAddress] != address(0),
-            "Committee does not exist."
+            _metachainId != bytes32(0),
+            "Metachain id is 0."
         );
 
         require(
-            !reputation.isSlashed(_validator),
-            "Validator is slashed."
+            _core != CoreI(0),
+            "Core's address is 0."
         );
 
-        CommitteeI committee = CommitteeI(_committeeAddress);
+        require(
+            _validator != address(0),
+            "Validator's address is 0."
+        );
+
+        require(
+            _furtherMember != address(0),
+            "Further validator's address is 0."
+        );
+
+        require(
```

Comment:
can we collapse these into a modifier 
```
onlyAssignedCore(_metachainId)
```
in a new ticket?

Reply:
In the current implementation of `enterCommittee` the core of the validator is specified by its address (not metablockchain id). In upcoming implementation of `enterCommittee` (according to the newest proposal) we can.